### PR TITLE
Focal-MSE loss for AirfRANS volume — upweight hard vol predictions

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Volume predictions have high variance across cells — easy regions (freestream) dominate the MSE while hard regions (near-wall, wake, separation zones) contribute relatively little to the gradient signal. Focal loss (Lin et al., 2017, https://arxiv.org/abs/1708.02002) addresses this by upweighting hard examples: `focal_mse = (error / error_max)^gamma * error^2`. Applied to the volume loss, this should push the model to reduce errors in the high-error cells that matter most for physical accuracy, without touching the surface loss at all.

The gamma parameter controls the strength of the upweighting:
- **gamma=2** (Trial 1): strong focus on hardest predictions — similar to the original focal loss for classification
- **gamma=1** (Trial 2): moderate / linear upweighting — softer version, less aggressive

We keep vol-loss-weight=10x (the current champion config) so the absolute scale of the volume loss is unchanged. The focal weighting only redistributes attention *within* the volume loss toward harder cells.

## Instructions

**Implementation (5 steps):**

1. Compute per-point squared error for volume predictions:
   ```python
   per_point_mse = (pred_vol - target_vol) ** 2  # shape: [N, C]
   ```

2. Normalize by the max error (detached from gradient graph):
   ```python
   error_max = per_point_mse.detach().max().clamp(min=1e-8)
   normalized = per_point_mse.detach() / error_max  # shape: [N, C], in [0, 1]
   ```

3. Compute the focal weight (also detached):
   ```python
   focal_weight = normalized ** gamma  # gamma is a hyperparameter
   ```

4. Apply the focal weight to the actual (non-detached) per-point MSE:
   ```python
   focal_mse = focal_weight * per_point_mse  # gradient flows through per_point_mse only
   ```

5. Use the mean as the volume loss, with vol-loss-weight=10x:
   ```python
   vol_loss = focal_mse.mean()
   ```

**Critical:** `focal_weight` must be detached. The gradient should only flow through `per_point_mse`, not through the normalisation. This is the same trick used in focal loss for detection.

---

**Smoke test first (3 epochs)** to verify no NaN / shape errors before launching full runs.

---

**Trial 1 — gamma=2 (strong focus):**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --vol-loss-weight 10.0 \
  --focal-vol-gamma 2.0 \
  --wandb_group af-focal-volume-loss
```

**Trial 2 — gamma=1 (moderate/linear upweighting):**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --vol-loss-weight 10.0 \
  --focal-vol-gamma 1.0 \
  --wandb_group af-focal-volume-loss
```

**Note:** If `--focal-vol-gamma` is not yet a CLI flag, add it to `train.py` (argparse + plumb into the volume loss computation). The implementation should be entirely within the loss function — no changes to the model architecture.

Report `val_primary/surface_mse` and `val_primary/vol_mse` for both trials. The target is `vol_mse < 0.0017` with no regression on `surface_mse` (must stay <= 0.000296).

## Baseline

Current best AirfRANS metrics (PR #3135, nezuko/af-ema-vol-weight-10x):

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_mse | **0.000296** | 704 |
| val_primary/vol_mse | **0.002039** | 743 |

- Baseline W&B run: `sh2zyfwr` (https://wandb.ai/run/sh2zyfwr)
- Config: 2L/256d/4H, AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, vol-weight=10x, Fourier

Reproduce command:
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --vol-loss-weight 10.0
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT casca:

## Results

### Trial 1 — gamma=2 (strong focal focus)

| Metric | Baseline | This run | Delta |
|--------|----------|----------|-------|
| val_primary/surface_mse | **0.000296** (ep704) | 0.001456 (ep198) | +392% ❌ |
| full_val/volume_mse | **0.002039** (ep743) | 0.08343 (ep198) | +3993% ❌ |
| test_primary/surface_mse | — | 0.003835 | — |
| full_test/volume_mse | — | 0.09344 | — |

- **W&B run:** [332df4z1](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/332df4z1)
- **Total epochs:** ~763 (timeout)
- **Best epoch:** 198 — model regressed dramatically after this point
- **Peak memory:** ~2.5 GB VRAM

**Command:**
\`\`\`bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 --focal-vol-gamma 2.0 \
  --agent casca --wandb-name "casca/af-focal-vol-gamma2" --wandb-group af-focal-volume-loss
\`\`\`

### Trial 2 — gamma=1 (moderate/linear upweighting)

| Metric | Baseline | This run | Delta |
|--------|----------|----------|-------|
| val_primary/surface_mse | **0.000296** (ep704) | 0.000952 (ep585) | +222% ❌ |
| full_val/volume_mse | **0.002039** (ep743) | 0.008557 (ep585) | +320% ❌ |
| test_primary/surface_mse | — | 0.001651 | — |
| full_test/volume_mse | — | 0.01008 | — |

- **W&B run:** [kasxtssl](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/kasxtssl)
- **Total epochs:** ~762 (timeout)
- **Best epoch:** 585
- **Peak memory:** ~2.5 GB VRAM

**Command:**
\`\`\`bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 --focal-vol-gamma 1.0 \
  --agent casca --wandb-name "casca/af-focal-vol-gamma1" --wandb-group af-focal-volume-loss
\`\`\`

### What happened

**Both trials regressed significantly on both surface and volume metrics.** The focal-MSE loss hurt rather than helped.

- **gamma=2 (strong focus)** was catastrophic — best surface_mse was 4.9x worse than baseline and vol_mse was 41x worse. The model peaked early (ep198) and never recovered, suggesting the aggressive focal reweighting destabilized the volume loss landscape.
- **gamma=1 (moderate)** performed better than gamma=2 but still substantially worse than baseline on both metrics: surface +222%, volume +320%. Best epoch was 585 (vs baseline 704/743), suggesting training was converging more slowly.

**Why it failed:** The focal weighting upweights high-error predictions, which in early training are near-wall/wake cells. But these are also the cells where the model needs the most gradient signal to learn the physics correctly. The focal weight suppresses the easy freestream gradients that help stabilize the optimization trajectory — essentially cutting off the "scaffolding" gradients that the model needs to learn the global flow structure before refining local details. The vol-loss-weight=10x baseline already provides strong volume gradient signal; the focal redistribution just makes the gradient noisier without improving the signal-to-noise ratio.

The stronger gamma (2) was much worse because it more aggressively suppresses the stabilizing freestream gradients, leading to early divergence (ep198 best vs ep585 for gamma=1).

### Suggested follow-ups

- **Curriculum focal scheduling** — start with gamma=0 (standard MSE) for the first N epochs to let the model learn global structure, then ramp gamma up to refocus on hard cells. This could get the best of both worlds.
- **Inverse focal** (gamma < 0) — upweight easy cells to stabilize training even further. Though this seems unlikely to beat the current baseline.
- **Spatial proximity weighting** instead of error-based weighting — weight volume cells by their distance to the surface, since near-surface cells are physically more important regardless of current error magnitude.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3227 because ADVISOR: Closing — focal-MSE is a clear dead end for AF volume. gamma=2: surface +392%, vol +3991% (catastrophic). gamma=1: surface +222%, vol +320%. Root cause: the batch-max normalization makes the loss landscape non-stationary — the denominator shifts every iteration as predictions improve, destabilizing gradient signal. Also, freestream cells that get downweighted serve as essential 'scaffolding' for the optimization trajectory. Your mechanistic analysis was excellent. The curriculum focal idea is interesting but needs a ground-up reformulation (e.g., EMA-smoothed normalization) rather than iteration on this approach..
